### ISSUE TITLE
Suggestion: add isNaN check to fromNullable

### ISSUE THREAD

**Issue by raveclassic** [gcanti/fp-ts#309]
What about adding `isNaN` check to `Option`'s `fromNullable` for `number` input? Seems like there's no profit from `NaN`-values, only pain.

**Comment by SimonMeskens** [comment]
Actually, NaN is the Maybe monad. I always really liked it. Makes sense to add NaN support to Option though.

**Comment by raveclassic** [comment]
It's quite confusing every time your QA team reports there's some sudden "NaN" string somewhere in the UI. And there's no help from TS.

**Comment by SimonMeskens** [comment]
The issue is that right now, you pass the number monad (let's call it that) all the way to the UI. At that point it's confusing to know where the failure came from. Let's say you replace it with Option<number> and pass it all the way to the UI. Same issue. Option doesn't actually fix this problem. What you want to do is at some point, check the Option to see if it's nil or not, but you might just as well just check the number for NaN at that point. NaN is for all intents and purposes a fully functioning, law-abiding Maybe monad.

That said, the issue is that there's no equivalent of Just<number> for the number monad. There's no way to explicitly say: I checked this number and it's not NaN, or: I only want to accept real numbers, not NaN. What we need is a way to type that, so we can figure out the part of the code where we want to be sure it's a number and properly statically type that part to only accept numbers.

I'm not sure what the proper solution here is, but:
- It's not just to add isNaN support to fromNullable
- Like with nullables, Option should indeed be able to lift from number to Option

Does that make sense?

**Comment by sledorze** [comment]
As a number can become NaN after some operations, maybe it would be acceptable to have a kind of opaque type (newtype like), a function converting to None if NaN and otherwise to Some(..).

**Comment by raveclassic** [comment]
@SimonMeskens Totally makes sense. But it also means that every time you render a numeric you make that check, quite annoying. Maybe @sledorze's suggestion about newtypes would help, need to investigate.

Still I understand that checking isNaN in fromNullable does not solve the problem entirely but only a small part of it and I agree that Option should be able to process invalid numbers (maybe also invalid dates). Should it be the same old fromNullable or smth new?

**Comment by gcanti** [comment]
My 2 cents.

In TypeScript, `undefined`, `null` and `number` are different types, but `NaN` is a _value_ and _is_ a `number`.

Other than `NaN` there's `+/- Infinity`, they are all `number`s.

From a categorical perspective, if something is not encoded in types (objects), then it doesn't "exist" (meaning is not modelled).

Also `fromNullable :: A | null | undefined -> Option<A>` is polymorphic, its implementation is not supposed to inspect `A`.

What you want is a refinement of `number` (<= here I mean the actual TypeScript type, not an abstract concept of number).
Same for integers, positive numbers, positive integers, etc...

As @sledorze said newtypes can help here, they allow to define _types_ that represents refinements.

It's true that `NaN` could be considered the `None` of `number`s, but the same reasoning apply for any 2-partition of a set

```ts
// Given a set S and 2 partitions A U B = S
const fromS = (s: S): Option<A> => if s in A then Some(a) else None 
```

Basically you are defining a `Prism<S, A>`.

The point is: how you get `Infinity` in the first place? Two possible reasons

- (1) at the system's border
- (2) not well-defined operations (like divide by zero)

Solutions

- (1) input validation
- (2) change the codomain to `Option` or use newtypes

**Comment by raveclassic** [comment]
@gcanti Makes sense. So I ended up with something like this:
```typescript
import { Newtype, iso, getRing } from 'newtype-ts'
import { fieldNumber } from 'fp-ts/lib/Field'
import { Option, none, some } from 'fp-ts/lib/Option'
import { Ring } from 'fp-ts/lib/Ring'

type Real = Newtype<'Real', number>
const isoReal = iso<Real>()

const fromNumber = (n: number): Option<Real> => {
  if (isNaN(n) || n === Infinity || n === -Infinity) {
    return none as any
  }
  return some(isoReal.wrap(n))
}

const div = (a: Real) => (b: Real) => fromNumber(isoReal.unwrap(a) / isoReal.unwrap(b))
const mod = (a: Real) => (b: Real) => fromNumber(isoReal.unwrap(a) % isoReal.unwrap(b))

type RealOps = {
  div: typeof div
  mod: typeof mod
}
const Real: Ring<Real> & RealOps = {
  ...getRing(fieldNumber),
  div,
  mod
}

const sum = (xs: Real[]): Real => xs.reduce((acc, x) => Real.add(acc)(x), Real.zero())
const average = (xs: Real[]): Option<Real> => fromNumber(xs.length).chain(length => Real.div(sum(xs))(length))
```

Is there any room for improvements?

**Comment by gcanti** [comment]
> Is there any room for improvements?

Well, it depends on how far you are willing to go. 
In practice even `sum(a)(b)` can fail if `a` and/or `b` are big enough (`Number.MAX_VALUE`)

p.s.
Take a look at [money-ts](https://github.com/gcanti/money-ts)

**Comment by SimonMeskens** [comment]
@gcanti The difference between any 2-partition of a set is that NaN does error propagation. Mathematically speaking, NaN is an absorbing element for all operations on number, regardless of which operations have NaN as a result. I'm not sure, but this a somewhat stronger concept I think? I'm unsure on how to properly express it.

**Comment by raveclassic** [comment]
Seems like this issue is out of scope for `Option`'s `fromNullable` and can be closed.
